File: problems/0/1/1/011.py
# Copyright (C) 2013 Jorge Aparicio

directions = [(1, 0), (0, 1), (1, 1), (1, -1)]
size = 20
stride = 4

with open('011.in') as f:
    arr = list(map(lambda line: list(map(int, line.split())), f.read().split('\n')))

ans = 0
for direction in directions:
    for i in range(size):
        if not 0 <= i + (stride - 1) * direction[0] < size:
            continue
        for j in range(size):
            if not 0 <= j + (stride - 1) * direction[1] < size:
                continue
            p = arr[i][j]
            for k in range(1, stride):
                p *= arr[i + k * direction[0]][j + k * direction[1]]

            if p > ans:
                ans = p

print(ans)

File: problems/0/1/1/011.rs
// Copyright (C) 2014 Jorge Aparicio

use std::io::fs::File;
use std::vec_ng::Vec;

static STRIDE: int = 4;

fn main() {
    let content = File::open(&Path::new("011.in")).read_to_str().unwrap();

    let grid: Vec<Vec<uint>> =
        content.
        lines().
        map(|line| line.words().filter_map(from_str).collect()).
        collect();

    let (nr, nc) = (grid.len() as int, grid.get(0).len() as int);
    let directions = [(0, 1), (1, 0), (1, 1), (-1, 1)];

    let mut max = 0;
    for &(y_step, x_step) in directions.iter() {
        for r in range(0, nr) {
            let y_max = r + STRIDE * y_step;
            if y_max > nr || y_max < -1 { continue }

            for c in range(0, nc) {
                let x_max = c + STRIDE * x_step;
                if x_max > nc || x_max < -1 { continue }

                let mut tmp = *grid.get(r as uint).get(c as uint);
                for i in range(1, STRIDE) {
                    let x = (c + i * x_step) as uint;
                    let y = (r + i * y_step) as uint;

                    tmp *= *grid.get(y).get(x);
                }

                if tmp > max { max = tmp }
            }
        }
    }

    println!("{}", max);
}

